threaddb/buckets.go

```diff
@@ -105,6 +300,138 @@ func init() {
+			}
+			var type = event.patch.type
+			var patch = event.patch.json_patch
+			var restricted = ["owner", "name", "version", "key", "archives", "created_at"]
+			switch (type) {
+			  case "create":
+				if (patch.owner !== "" && writer !== patch.owner) {
+				  return "permission denied" // writer must match new bucket owner
+				}
+				break
+			  case "save":
+				if (instance.owner === "") {
+				  return true
+				}
+				if (writer !== instance.owner) {
+				  for (i = 0; i < restricted.length; i++) {
+					if (patch[restricted[i]]) {
+					  return "permission denied"
+					}
+				  }
+				}
+				if (!patch.metadata) {
+			      if (patch.path && writer !== instance.owner) {
+			        return "permission denied"
+			      } else {
+			        patch.metadata = {}
+			      }
+				}
+				var keys = Object.keys(patch.metadata)
+				for (i = 0; i < keys.length; i++) {
+				  var p = patch.metadata[keys[i]]
+				  var x = instance.metadata[keys[i]]
+				  if (x) {
+					if (!x.roles[writer]) {
+					  x.roles[writer] = 0
+					}
+					if (!x.roles["*"]) {
+					  x.roles["*"] = 0
+					}
+					// merge all parents, taking most privileged role
+					if (keys[i].length > 0) {
+					  var parts = keys[i].split("/")
+					  parts.unshift("")
+					  var path = ""
+					  for (j = 0; j < parts.length; j++) {
+						if (path.length > 0) {
+						  path += "/"
+						}
+						path += parts[j]
+						var y = instance.metadata[path]
+						if (!y) {
+						  continue
+						}
+						if (!y.roles[writer]) {
+						  y.roles[writer] = 0
+						}
+						if (!y.roles["*"]) {
+						  y.roles["*"] = 0
+						}
+						if (y.roles[writer] > x.roles[writer]) {
+						  x.roles[writer] = y.roles[writer]
+						}
+						if (y.roles["*"] > x.roles["*"]) {
+						  x.roles["*"] = y.roles["*"]
+						}
+					  }
+					}
+					// check access against merged roles
+					if (!p) {
+			          if (x.roles[writer] < 3) {
+			            return "permission denied" // no admin access to delete items
+			          }
+			        } else {
+			          if (p.roles && x.roles[writer] < 3) {
+			            return "permission denied" // no admin access to edit roles
+			          }
+			        }
+			        if (x.roles[writer] < 2 && x.roles["*"] < 2) {
+			          return "permission denied" // no write access
+			        }
+			      } else {
+			        if (writer !== instance.owner) {
+			          return "permission denied" // no owner access to create items
+			        }
+			      }
+			    }
+			    break
+			  case "delete":
+			    if (instance.owner !== "" && writer !== instance.owner) {
+			      return "permission denied" // no owner access to delete instance
+			    }
+			    break
+			}
+			return true
+		`,
+		ReadFilter: `
+			if (!instance.version) {
+			  return instance
+			}
+			if (instance.owner === "") {
+			  return instance
+			}
+			var filtered = {}
+			var keys = Object.keys(instance.metadata)
+			outer: for (i = 0; i < keys.length; i++) {
```

Comment:
The idea here is to walk each metadata object, collecting ones that the identity has access to, considering the tree inheritance pattern mentioned earlier.